Magmalte rock 1.8

saltiyazan opened the pull request:
Builds on[ PR #1  ](https://github.com/canonical/magma-nms-magmalte-rock/pull/1) to fix the rock and to use magma 1.8

gruyaume committed:
Draft for the nms rock

saltiyazan committed:
Uses 1.8

saltiyazan committed:
Fixes yarn installation rockcraft file

saltiyazan committed:
Fixes tag references

gruyaume commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,28 @@
+        run: |
+          sudo snap install --classic --channel edge rockcraft
+      - name: Build ROCK
+        run: rockcraft pack --verbose
```
Please use the `canonical/craft-actions/rockcraft-pack@main` workflow. You can see an example [here](https://github.com/canonical/magma-orc8r-controller-rock/blob/main/.github/workflows/build.yaml)

gruyaume replied:
overall you can very much re-use the workflows from the project linked above (also applicable to `push.yaml`)

saltiyazan replied:
Done

gruyaume commented on README.md:
````diff
@@ -1 +1,14 @@
+## Usage
+
+```bash
+docker pull ghcr.io/gruyaume/magma-orc8r-nms-magmalte:1.8.0
````
```suggestion
docker pull ghcr.io/canonical/magma-orc8r-nms-magmalte:1.8.0
```

saltiyazan replied:
✅

gruyaume commented on README.md:
````diff
@@ -1 +1,14 @@
+
+```bash
+docker pull ghcr.io/gruyaume/magma-orc8r-nms-magmalte:1.8.0
+docker run -it ghcr.io/gruyaume/magma-orc8r-nms-magmalte:1.8.0
````
```suggestion
docker run -it ghcr.io/canonical/magma-orc8r-nms-magmalte:1.8.0
```

saltiyazan replied:
✅

gruyaume commented on README.md:
```diff
@@ -1 +1,14 @@
\ No newline at end of file
+# Magma Orchestrator NMS Rock
+
+> **Warning**: This rock is under construction
```
Remove this line

saltiyazan replied:
✅

gruyaume commented on rockcraft.yaml:
```diff
@@ -0,0 +1,59 @@
+    override-build: |
+      mkdir -p $CRAFT_PART_INSTALL/bin/
+      make defconfig
+      make
```
Isn't there a more `rockcraft` way to do this part? Meaning leveraging one of the plugins and avoiding `override-build`. I know I wrote it but it's not super beautiful

ghislainbourgeois replied:
I just looked at the documentation for the `make` plugin, and it basically says to use `override-build` for what we need to do: https://snapcraft.io/docs/make-plugin

I think in that case, using the plugin brings no value.

saltiyazan replied:
I would have tested with the [kbuild](https://snapcraft.io/docs/kbuild-plugin) plugin from Snapcraft, but Rockcraft doesn't have the same at the moment.

gruyaume commented on rockcraft.yaml:
```diff
@@ -0,0 +1,59 @@
+
+    override-stage: |
+      set -x
+      curl -s https://nodejs.org/dist/v16.14.1/node-v16.14.1-linux-x64.tar.gz | tar xzf - -C ${CRAFT_PART_INSTALL}/ --no-same-owner --strip-components=1
```
Maybe we should have a stage just to install yarn. I don't know if it needs to be in this one. Or maybe not, I'm not sure

saltiyazan replied:
I don't exactly get what you mean.
But I realised that we don't need to have node installed twice, so I removed the snap and moved this line to the build part.

saltiyazan committed:
Addresses review comments

ghislainbourgeois commented on README.md:
```diff
@@ -1 +1,12 @@
+
+Rock for Magma's orchestrator NMS service built using 
+[rockcraft](https://github.com/canonical/rockcraft). It provides a direct port of this 
+[Dockerfile](https://github.com/magma/magma/blob/v1.6/nms/app/packages/magmalte/Dockerfile).
```
Should this point to 1.8?
```suggestion
[Dockerfile](https://github.com/magma/magma/blob/v1.8/nms/app/packages/magmalte/Dockerfile).
```

saltiyazan replied:
✅

Gmerold commented on .github/workflows/push.yaml:
```diff
@@ -0,0 +1,60 @@
+    paths:
+      - "rockcraft.yaml"
+      - "files/**"
+      - ".github/workflows/**.yaml"
```
Do we care to push out the new rock when workflows change?

saltiyazan replied:
Fixed this and added new workflow files.

Gmerold commented on .github/workflows/push.yaml:
```diff
@@ -0,0 +1,60 @@
+
+      - uses: actions/download-artifact@v3
+        with:
+          name: magma-orc8r-nms-magmalte-rock
```
Artifact's name, as defined in `build.yaml`, is `rock` not `magma-orc8r-nms-magmalte-rock.`

saltiyazan replied:
fixed

Gmerold commented on README.md:
```diff
@@ -1 +1,12 @@
\ No newline at end of file
+# Magma Orchestrator NMS Rock
+
+Rock for Magma's orchestrator NMS service built using 
```
```suggestion
Rock for Magma Orchestrator's NMS service built using 
```
?

saltiyazan replied:
✅

saltiyazan committed:
Improves the project and adds integration tests

saltiyazan committed:
Uses ubuntu 22.04

Gmerold commented on tests/integration/test_integration.py:
```diff
@@ -0,0 +1,89 @@
+
+import docker  # type: ignore[import]
+import yaml
+import requests  # type: ignore[import]
```
You could avoid this ignore by adding `types-requests` to the `tox.ini`.

saltiyazan replied:
✅

Gmerold commented on tests/tox.ini:
```diff
@@ -0,0 +1,75 @@
+    isort {[vars]integration_test_path}
+    black {[vars]integration_test_path}
+
+[testenv:lint]
```
Should we also add `lint` and `static` to the CI since it's already here?

saltiyazan replied:
It would cover the integration test only, should we add it?

Gmerold replied:
We can either add it as additional checks or just remove the relevant sections from this file.

saltiyazan replied:
Added the checks :)

saltiyazan commented:
[Issue](https://github.com/canonical/craft-actions/issues/6) created on Craft-actions to address the failure in the ci.

saltiyazan committed:
Adds lint and static analysis to the ci

saltiyazan committed:
Minor improvements

saltiyazan committed:
Fixes port binding in integration test

saltiyazan committed:
Fixes lint issue

saltiyazan committed:
Fixes integration test

saltiyazan committed:
Debugging integration test

gruyaume commented on tests/integration/test_integration.py:
```diff
@@ -0,0 +1,94 @@
+import requests
+import yaml
+
+#MAGMALTE_DOCKER_URL = "http://172.17.0.3"
```
:cry:

saltiyazan replied:
debugging github runner in progress :)

saltiyazan replied:
it should be all good now :)

saltiyazan committed:
Installs nodejs

ghislainbourgeois commented on .github/workflows/lint-report.yaml:
```diff
@@ -0,0 +1,14 @@
+  build:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v2
```
In another workflow we use v3, we should probably standardize.
```suggestion
      - uses: actions/checkout@v3
```

saltiyazan replied:
✅

ghislainbourgeois commented on .github/workflows/static-analysis.yaml:
```diff
@@ -0,0 +1,14 @@
+  build:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v3
```

saltiyazan replied:
✅

ghislainbourgeois commented on rockcraft.yaml:
```diff
@@ -0,0 +1,59 @@
+      ln -s ${CRAFT_PART_INSTALL}/node_modules/yarn/bin/yarn ${CRAFT_PART_INSTALL}/usr/local/bin/yarn
+      craftctl default
+
+  busybox:
+    plugin: nil
+    source: https://busybox.net/downloads/busybox-1.34.1.tar.bz2
+    build-packages:
+      - make
+      - gcc
+    stage-packages:
+      - libc6
+    override-build: |
+      mkdir -p $CRAFT_PART_INSTALL/bin/
+      make defconfig
+      make
+      mv busybox $CRAFT_PART_INSTALL/bin/
```
Just thought about this, but `busybox` is in the Ubuntu repos, can't we use that package directly? We would not have to maintain this whole block.

saltiyazan replied:
Really good catch, thanks.
Done

ghislainbourgeois commented on tests/integration/test_integration.py:
```diff
@@ -0,0 +1,91 @@
+
+    def setUp(self) -> None:
+        """Run containers to test."""
+        super().setUp()
```
This line does nothing.
```suggestion
```

saltiyazan replied:
👍

ghislainbourgeois commented on tests/integration/test_integration.py:
```diff
@@ -0,0 +1,91 @@
+        self,
+    ):
+        """Test to validate that the container is running correctly."""
+        sleep(10)
+        response = requests.get(
+            f"{MAGMALTE_DOCKER_URL}:{MAGMALTE_DOCKER_PORT}{MAGMALTE_LOGIN_PAGE}"  # noqa: E501
+        )
```
```suggestion
        for _ in range(10):
            response = requests.get(
                f"{MAGMALTE_DOCKER_URL}:{MAGMALTE_DOCKER_PORT}{MAGMALTE_LOGIN_PAGE}"  # noqa: E501
            )
            sleep(1)
```

saltiyazan replied:
Done.

saltiyazan committed:
Addresses review comments
Installs busybox in stage-packages
Uses v3 of checkout action
Waits for docker container to be up in a loop

ghislainbourgeois commented on tests/integration/test_integration.py:
```diff
@@ -0,0 +1,96 @@
+    ):
+        """Test to validate that the container is running correctly."""
+        url = f"{MAGMALTE_DOCKER_URL}:{MAGMALTE_DOCKER_PORT}{MAGMALTE_LOGIN_PAGE}"  # noqa: E501
+        for _ in range(60):
```
This loop might be a bit long, it will basically retry for a minute.

saltiyazan replied:
I increased the time just to be sure not to fail the test if the pod takes longer than expected to start and run the service.

saltiyazan replied:
At the end I decreased it to 30 :)

ghislainbourgeois reviewed (approved):
Small nitpick on my side around the retry loop in the test, otherwise LGTM

saltiyazan committed:
Decreases waiting time in integration test

sanchezfdezjavier commented on tests/tox.ini:
```diff
@@ -0,0 +1,76 @@
+    -r{toxinidir}/requirements.txt
+    mypy
+    types-PyYAML
+    pytest
+    pytest-operator
+    juju
```
Same here?

sanchezfdezjavier commented on tests/tox.ini:
```diff
@@ -0,0 +1,76 @@
+[testenv]
+deps =
+    pytest
+    pytest-operator
```
Do we need this guy?

saltiyazan replied:
✅

sanchezfdezjavier commented on tests/tox.ini:
```diff
@@ -0,0 +1,76 @@
+    mypy
+    types-PyYAML
+    pytest
+    pytest-operator
+    juju
```
Same here.

saltiyazan replied:
✅

sanchezfdezjavier commented on tests/pyproject.toml:
```diff
@@ -0,0 +1,52 @@
+ignore_missing_imports = true
+
+[[tool.mypy.overrides]]
+module = ["charms.*"]
```
I think we're not using this guy.

saltiyazan replied:
✅

sanchezfdezjavier commented on tests/pyproject.toml:
```diff
@@ -0,0 +1,52 @@
+
+# Ignore libraries that do not have type hint nor stubs
+[[tool.mypy.overrides]]
+module = ["ops.*", "kubernetes.*", "flatten_json.*", "git.*"]
```
I think we're not using this guys.

saltiyazan replied:
✅

saltiyazan committed:
Removes unused dependencies from tox.ini

sanchezfdezjavier commented on tests/integration/test_integration.py:
```diff
@@ -0,0 +1,96 @@
+POSTGRES_USER = "username"
+POSTGRES_PASSWORD = "password"
+POSTGRES_DB = "magma"
+DATABASE_SOURCE = f"""
+    dbname=magma
+    user={POSTGRES_USER}
+    password={POSTGRES_PASSWORD}
+    host=postgres_container
+    sslmode=disable
+"""
```
Are we using this guy?

saltiyazan replied:
no, not anymore.
✅

saltiyazan committed:
Addresses review comments

saltiyazan committed:
Uses only one file for push workflow

sanchezfdezjavier reviewed (commented):
LGTM
